### DISCUSSION TITLE
Is it possible to build a source distribution archives [ i.e creating tar.gz file] excluding particular file

### DISCUSSION THREAD

**Discussion by Earammak** [pypa/cibuildwheel#1548] [General]
I have a requirement to build a source distribution package to exclude particular file and create a tar.gz file.
Is there any feature to exclude and create a source distribution.

Thanks

**Comment by Czaki** [comment]
You could use MANIFEST.in file for this https://packaging.python.org/en/latest/guides/using-manifest-in/

**Comment by Earammak** [comment]
Thank you for quick response.
When we have file structure in a way that the directory contains multiple sub directories. I want to delete complete directory.

This directory needs to get downloaded from setup.py, when we are installing package based on the platform.
So, can you please help how we could exclude the directory.


Thanks

**Reply by Czaki** [reply]
Please red the documentation. It looks like `recursive-exclude dir-pattern pat1 pat2` should be good for you.

**Reply by henryiii** [reply]
> when we are installing package based on the platform

There is exactly 1 SDist. You can't put different things in the SDist based on the platform. You can do this with wheels, but then you'll need to handle downloading the files in the wheel building step in your setup.py.

**Comment by Earammak** [comment]
I tried with that actually, but after installing the package from that source distribution under site-packages, those subdirectories are not being added. 
Those sub-directories are supposed to be there for my package.

**Reply by Czaki** [reply]
if you do not add subdirectory to s dist then it is your responsibility to write code in setup.py that will download it.

**Reply by Earammak** [reply]
That file is tar.gz file, while installing packages based on the platform it should pick the tar.gz file, download and untar the file. 
When the file downloaded means, it is supposed to have all files. 
This is what my requirement.

**Reply by henryiii** [reply]
It's not clear to me if you want to exclude a directory (`prune <directory>`) or add one (`graft <directory>`), but both things can be done in `MANIFEST.in`. Syntax of the file is here: https://packaging.python.org/en/latest/guides/using-manifest-in/#manifest-in-commands

It only affects Source -> SDist, though. Not SDist -> wheel, if you are looking for that.

**Reply by Earammak** [reply]
My concern here is that , we can build source distribution by using one of the way from MANIFEst.in file. 
When we are installing a package, our setpu.py needed to download file.
In install log we see that file downloading and after this MANIFEST file is running. 
Since we excluded in MANIFEST.in file,  during install it is not being included. 

I want a scenario in a way that  in source distribution file should not be included and while installing a package file has to be added. 
I'm not sure here how to approach this process.

**Reply by henryiii** [reply]
MANIFEST.in only affects SDist. Not wheel (which means not install)[^1]. You can use package_data (in setup.py, setup.cfg, or inside tool.setuptools in pyproject.toml) to control what setuptools puts into the wheel.

[^1]: if you set include_package_data to true, files allowed by the manifest will also be added to the wheel. That’s just a shortcut, though, and not helpful in your case.

**Reply by Earammak** [reply]
Example in my MANIFEST.in file i have added " recursive-exclude clidriver *". After adding this excluding clidriver source distribution created. But, in wheels file also it is excluded. 
Here is my package where i have released in testpypi. 
https://test.pypi.org/project/ibm-db/3.1.4.5/

Below is my requirement: 
Currently consider we are building wheels from python version 3.6 - 3.11 for operating systems linux, macos and windows.
When this requirement does not match, the package should get installed from source distribution.

Whatever file i was mentioned to exclude is to download based on the platform we use.
The process we do from github action, building source distribution runs on linux operating systems.
linux file will be included in source distribution. And i cannot install package successfully other than linux from source distribuion.

Example i want to install package on AIX machine, the file has to download for aix file only.
When i try to install package on this machine, it don't have corresponding file and not able to install successfully.

Please give some suggestion here to make it progress.

**Reply by Czaki** [reply]
Did I correctyl undersand that it works on linux machines, but does not work on another? Did you try to run `pip install --no-clean imb-db` and then inspect the content of the directory to check what is wrong?

**Reply by Earammak** [reply]
I know here the problem is because while installing package clidriver took for linux, which is not the right one to that particular machine. 
So my concern is if we create a source distribution by excluding clidriver, while installing package it can install by downloading the corresponding clidriver.

**Comment by Earammak** [comment]
@Czaki Can you please suggest here ?
